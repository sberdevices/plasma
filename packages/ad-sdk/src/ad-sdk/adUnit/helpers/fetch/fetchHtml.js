const isValidContentType = (contentType) => {
    const normalisedCT = contentType.toLowerCase();

    return ["text/plain", "text/html"].some((allowedType) => normalisedCT.includes(allowedType));
};

const fetchHtml = async (endpoint) => {
    const response = await fetch(endpoint);
    const contentType = response.headers.get("Content-Type");

    if (response.status >= 400) {
        const error = new Error(response.statusText);

        error.response = response;
        throw error;
    }

    if (!isValidContentType(contentType)) {
        const error = new Error(`fetchHtml error, invalid Content-Type ${contentType}`);

        error.response = response;
        throw error;
    }

    return response.text();
};

export default fetchHtml;
