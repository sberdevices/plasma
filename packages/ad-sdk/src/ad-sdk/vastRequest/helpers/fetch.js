const fetch = async (endpoint, options = {}) => {
    const defaults = {
        credentials: "include",
    };
    const fetchOptions = Object.assign({}, defaults, options);
    const response = await window.fetch(endpoint, fetchOptions);

    if (response.status >= 400) {
        const error = new Error(response.statusText);

        error.response = response;
        throw error;
    }

    return response;
};

export default fetch;
