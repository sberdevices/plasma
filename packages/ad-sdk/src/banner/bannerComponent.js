function createCrossIcon({ onClickCloseButton }) {
    const crossCircleIcon = document.createElementNS("http://www.w3.org/2000/svg", "svg");
    crossCircleIcon.style.display = "block";
    crossCircleIcon.setAttribute("width", "40px");
    crossCircleIcon.setAttribute("height", "40px");
    crossCircleIcon.setAttribute("viewBox", "0 0 24 24");
    crossCircleIcon.setAttribute("fill", "none");
    crossCircleIcon.setAttribute("color", "inherit");

    const crossCircleIconContent = document.createElementNS("http://www.w3.org/2000/svg", "path");
    crossCircleIconContent.setAttribute(
        "d",
        "M12 2c5.522 0 10 4.477 10 10s-4.478 10-10 10S2 17.523 2 12 6.478 2 12 2zM8.707 7.292a1 1 0 00-1.414 1.415L10.586 12l-3.293 3.293a.999.999 0 00-.083 1.32l.083.094a.997.997 0 001.414 0L12 13.414l3.293 3.293a.996.996 0 00.58.285L16 17a.999.999 0 00.707-1.707L13.414 12l3.293-3.293a1 1 0 00.083-1.32l-.083-.095a1 1 0 00-1.414 0L12 10.585 8.707 7.292z"
    );
    crossCircleIconContent.setAttribute("fill", "white");
    crossCircleIcon.appendChild(crossCircleIconContent);
    const button = document.createElement("button");
    button.appendChild(crossCircleIcon);
    button.style.background = "none";
    button.style.border = "none";
    button.style.padding = "0";
    button.style.margin = "0";
    button.style.position = "relative";
    button.addEventListener("click", (event) => {
        event.stopPropagation();
        onClickCloseButton(event);
    });
    return button;
}

function createBackBlock({ onClickCloseButton, isTvRemote }) {
    if (isTvRemote) {
        const backTitle = document.createElement("div");
        backTitle.style.color = "white";
        backTitle.innerHTML = 'Чтобы вернуться нажмите "назад"';
        return backTitle;
    } else {
        const crossCircleIcon = createCrossIcon({ onClickCloseButton });
        crossCircleIcon.style.flexGrow = "0";
        crossCircleIcon.style.flexShrink = "0";

        return crossCircleIcon;
    }
}

function createTimerText({ autoCloseTime, timerId }) {
    const timerText = document.createElement("div");
    timerText.innerHTML = `Реклама закроется через <span id="${timerId}">${autoCloseTime}</span>с`;
    timerText.style.color = "white";
    return timerText;
}

function getIframeSize(width, height) {
    let resultWidth;
    let resultHeight;
    if (width.indexOf("%") !== -1) {
        resultWidth = "100%";
    } else {
        resultWidth = width;
    }

    if (height.indexOf("%") !== -1) {
        resultHeight = "100%";
    } else {
        resultHeight = height;
    }

    return {
        width: resultWidth,
        height: resultHeight,
    };
}

function getBannerContainerSize(width, height) {
    let resultWidth;
    let resultHeight;
    if (width.indexOf("%") !== -1) {
        resultWidth = width;
    }

    if (height.indexOf("%") !== -1) {
        resultHeight = height;
    }

    return {
        width: resultWidth,
        height: resultHeight,
    };
}

export function createBanner({
    iframeUrl,
    width,
    height,
    timerId,
    autoCloseTime,
    onLoadIframe,
    onClickCloseButton,
    isTvRemote,
}) {
    const iframeSize = getIframeSize(width, height);
    const adIframe = document.createElement("iframe");
    adIframe.id = "plasma-ad-frame";
    adIframe.setAttribute("src", iframeUrl);
    adIframe.setAttribute("width", iframeSize.width);
    adIframe.setAttribute("height", iframeSize.height);
    adIframe.style.border = "none";
    adIframe.addEventListener("load", onLoadIframe);

    const adWrapper = document.createElement("div");
    adWrapper.style.position = "absolute";
    adWrapper.style.top = "0";
    adWrapper.style.left = "0";
    adWrapper.style.width = "100%";
    adWrapper.style.height = "100%";
    adWrapper.style.display = "none";

    const containerSize = getBannerContainerSize(width, height);
    const banner = document.createElement("div");
    banner.style.margin = "auto";
    banner.style.flexShrink = "0";
    banner.style.flexGrow = "0";
    banner.style.maxWidth = "100vw";
    banner.style.maxHeight = "100vh";
    banner.style.padding = "5px";
    banner.style.boxSizing = "border-box";
    banner.style.display = "flex";
    banner.style.flexDirection = "column";
    banner.style.alignItems = "strength";
    if (containerSize.width) {
        banner.style.width = containerSize.width;
    }
    if (containerSize.height) {
        banner.style.height = containerSize.height;
    }

    if (isTvRemote) {
        const adHeader = document.createElement("div");
        adHeader.style.display = "flex";
        adHeader.style.alignItems = "center";
        adHeader.style.boxSizing = "border-box";
        adHeader.style.padding = "5px";
        adHeader.style.flexGrow = "0";
        adHeader.style.flexShrink = "0";

        const backBlock = createBackBlock({ onClickCloseButton, isTvRemote });

        const textTimer = createTimerText({ autoCloseTime, timerId });
        textTimer.style.flexGrow = "1";
        textTimer.style.flexShrink = "1";

        adHeader.appendChild(textTimer);
        adHeader.appendChild(backBlock);

        banner.appendChild(adHeader);
    }

    if (isTvRemote) {
        const linkWrapper = document.createElement("a");
        linkWrapper.id = "plasma-ad-banner-link";
        linkWrapper.setAttribute("href", "https://www.google.com/");
        linkWrapper.appendChild(adIframe);
        linkWrapper.style.flexGrow = "1";
        linkWrapper.style.flexShrink = "1";
        banner.appendChild(linkWrapper);
    } else {
        banner.appendChild(adIframe);
    }

    adWrapper.appendChild(banner);
    return adWrapper;
}

export function createPreloader({ id }) {
    const adWrapper = document.createElement("div");
    adWrapper.style.position = "absolute";
    adWrapper.style.top = "0";
    adWrapper.style.left = "0";
    adWrapper.style.width = "100%";
    adWrapper.style.height = "100%";
    adWrapper.style.display = "flex";
    adWrapper.id = id;

    const preloader = document.createElement("div");
    preloader.innerText = "Loading...";
    preloader.style.margin = "auto";
    preloader.style.color = "white";

    adWrapper.appendChild(preloader);
    return adWrapper;
}
