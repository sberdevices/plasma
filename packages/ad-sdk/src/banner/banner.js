import fetch from "../utils/fetch";
import { createBanner, createPreloader } from "./bannerComponent";

const TIME_TO_AUTO_CLOSE_IN_SECONDS = 10;

function blockFocusHandle(e) {
    e.preventDefault();
}

export class Banner {
    timeToClose = TIME_TO_AUTO_CLOSE_IN_SECONDS;
    autoCloseInterval = null;
    preloaderElement = null;
    _timerElement = null;
    _isTvRemote = false;
    _focusOutsideBanner = null;

    constructor(container, events = {}, sspParams, { isTvRemote } = {}) {
        this.container = container;
        this.onSuccess = events.onSuccess || (() => {});
        this.onError = events.onError || (() => {});
        this.onAdReady = events.onAdReady || (() => {});
        this.sspParams = sspParams;
        this._isTvRemote = isTvRemote;
        this._onIframeMessage = this._onIframeMessage.bind(this);
    }

    async run() {
        this._focusOutsideBanner = document.activeElement;
        history.pushState(null, null, "#plasma-ad-banner");
        this._listenOnPopState();
        this._showPreloader();

        try {
            const { source, height, width } = await this._requestBanner();

            const banner = createBanner({
                iframeUrl: source,
                width,
                height,
                timerId: "plasma-ad-time-to-close",
                autoCloseTime: TIME_TO_AUTO_CLOSE_IN_SECONDS,
                isTvRemote: this._isTvRemote,
                onLoadIframe: () => {
                    this._showBanner(banner);
                    this._hidePreloader();
                    this.onAdReady();

                    this._listenPostMessages();

                    if (this._isTvRemote) {
                        this._startAutoCloseTimer(banner);
                        this._blockFocus();
                    }
                },
                onClickCloseButton: () => {
                    this._closeAdSuccess();
                },
            });

            this.container.appendChild(banner);

            document.body.append(this.container);
        } catch (err) {
            this._closeAdError();
            this._hidePreloader();
        }
    }

    _onIframeMessage(message) {
        const childWindow = document.getElementById("plasma-ad-frame").contentWindow;
        if (message.source !== childWindow) {
            return;
        }
        if (message.data && message.data.event === "close") {
            this._historyGoBack();
        }
    }

    _listenPostMessages() {
        window.addEventListener("message", this._onIframeMessage);
    }

    _removePostMessagesListeners() {
        window.removeEventListener("message", this._onIframeMessage);
    }

    _blockFocus() {
        if (this._isTvRemote) {
            const link = document.getElementById("plasma-ad-banner-link");
            link.focus();
        }
        document.addEventListener("navbeforefocus", blockFocusHandle);
    }

    _showBanner(banner) {
        banner.style.display = "flex";
    }

    _showPreloader() {
        this.preloaderElement = createPreloader({ id: "plasma-ad-preloader" });
        this.container.appendChild(this.preloaderElement);
    }

    _hidePreloader() {
        this.preloaderElement.remove();
    }

    _closeAdSuccess() {
        this.onSuccess();
        this._historyGoBack();
    }

    _closeAdError() {
        this.onError();
        this._historyGoBack();
    }

    _close() {
        if (this._focusOutsideBanner && this._focusOutsideBanner.focus) {
            this._focusOutsideBanner.focus();
        }
        this._removePostMessagesListeners();
        this.container.remove();
        clearInterval(this.autoCloseInterval);
        document.removeEventListener("navbeforefocus", blockFocusHandle);
    }

    _listenOnPopState() {
        const onPopStateHandler = () => {
            this._close();
            window.removeEventListener("onpopstate", onPopStateHandler);
        };
        window.addEventListener("popstate", onPopStateHandler);
    }

    _historyGoBack() {
        if (window.location.hash === "#plasma-ad-banner") {
            history.back();
        }
    }

    _getTimerElement() {
        if (!this._timerElement) {
            this._timerElement = document.getElementById("plasma-ad-time-to-close");
        }
        return this._timerElement;
    }

    _startAutoCloseTimer() {
        const timer = this._getTimerElement();
        this.autoCloseInterval = setInterval(() => {
            this.timeToClose--;
            if (this.timeToClose < 1) {
                clearInterval(this.autoCloseInterval);
                this._closeAdSuccess();
                return;
            }
            timer.innerText = `${this.timeToClose}`;
        }, 1000);
    }

    async _requestBanner() {
        const url = "https://ssp.rambler.ru/sberdevapp.jsp?" + new URLSearchParams(this.sspParams);
        const response = await fetch(url);
        const { source, height, width } = await response.json();
        if (!source || !height || !width) {
            throw new Error("Некорректный ответ сервера");
        }
        return {
            source,
            height,
            width,
        };
    }
}
