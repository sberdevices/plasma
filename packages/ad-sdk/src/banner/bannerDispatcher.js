import { Banner } from "./banner";
import { injectCss } from "../utils/injectCss";

let enableToOpenBanner = true;

function generateStyles({ rootId, outlineSize = "0.125rem", outlineOffset = outlineSize, outlineRadius = 0 }) {
    return `
        #${rootId} button:focus::before {
            content: '';

            position: absolute;
            top: -${outlineOffset};
            left: -${outlineOffset};
            right: -${outlineOffset};
            bottom: -${outlineOffset};

            display: block;
            box-sizing: content-box;

            border: ${outlineSize} solid transparent;
            border-radius: ${outlineRadius};

            transition: box-shadow 0.2s ease-in-out;

            pointer-events: none;

            box-shadow: 0 0 0 ${outlineSize} green;
        }

        #${rootId} iframe a:focus::before {
            content: '';

            position: absolute;
            top: -${outlineOffset};
            left: -${outlineOffset};
            right: -${outlineOffset};
            bottom: -${outlineOffset};

            display: block;
            box-sizing: content-box;

            border: ${outlineSize} solid transparent;
            border-radius: ${outlineRadius};

            transition: box-shadow 0.2s ease-in-out;

            pointer-events: none;

            box-shadow: 0 0 0 ${outlineSize} green;
        }
    `;
}

export function openBanner({ container, params, cooldownTime, events, isTvRemote }) {
    if (!container.id) {
        throw new Error("id у container обязателен");
    }
    injectCss(generateStyles({ rootId: container.id }));
    const isAlreadyOpen = !!document.getElementById(container.id);
    if (!isAlreadyOpen && enableToOpenBanner) {
        enableToOpenBanner = false;
        setTimeout(() => {
            enableToOpenBanner = true;
        }, cooldownTime);

        const banner = new Banner(container, events, params, { isTvRemote });
        banner.run();
    } else {
        events.onError();
    }
}
