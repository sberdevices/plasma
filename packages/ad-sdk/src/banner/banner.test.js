import { Banner } from "./banner";
import fetch from "../utils/fetch";
jest.mock("../utils/fetch");
jest.useFakeTimers();

fetch.mockImplementation(() => {
    return Promise.resolve({ json: () => ({ source: "url", height: "100px", width: "150px" }) });
});

function waitAnimationFrame() {
    return new Promise((resolve, reject) => {
        jest.useRealTimers();
        window.requestAnimationFrame(() => {
            jest.useFakeTimers();
            resolve();
        });
    });
}

function triggerOnLoadIframe() {
    const iframe = document.getElementsByTagName("iframe")[0];
    iframe.dispatchEvent(new Event("load"));
}

function createBanner({ events, params } = {}) {
    const adContainer = document.createElement("div");
    adContainer.id = "root";
    const banner = new Banner(
        adContainer,
        {
            onError: (err) => console.log("error", err),
            ...events,
        },
        {},
        {
            ...params,
        }
    );
    return banner;
}

describe("banner.js - Banner class", () => {
    beforeEach(() => {
        document.body.innerHTML = "";
    });

    afterEach(() => {
        jest.runAllTimers();
        jest.clearAllTimers();
    });

    test("Совпадает снапшот для устройств без пульта", async () => {
        const banner = createBanner({ params: { isTvRemote: false } });
        await banner.run();
        triggerOnLoadIframe();
        const bodyElement = document.getElementsByTagName("body")[0];
        expect(bodyElement).toMatchSnapshot();
    });

    test("Совпадает снапшот для устройств с пультом", async () => {
        const banner = createBanner({ params: { isTvRemote: true } });
        await banner.run();
        triggerOnLoadIframe();
        const bodyElement = document.getElementsByTagName("body")[0];
        expect(bodyElement).toMatchSnapshot();
    });

    test("Отображается прелоадер перед началом загрузки банера", async () => {
        const banner = createBanner();
        await banner.run();

        const preloaderElement = document.getElementById("plasma-ad-preloader");

        expect(preloaderElement).toBeDefined();
    });

    test("Не отображается прелоадер после загрузки банера", async () => {
        const banner = createBanner();
        await banner.run();
        let preloaderElement = document.getElementById("plasma-ad-preloader");
        expect(preloaderElement).toBeDefined();
        triggerOnLoadIframe();
        preloaderElement = document.getElementById("plasma-ad-preloader");
        expect(preloaderElement).toBeFalsy();
    });

    test("После загрузки отображается iframe с правильными параметрами", async () => {
        const banner = createBanner();
        await banner.run();

        const iframe = document.getElementsByTagName("iframe")[0];

        expect(iframe).toBeTruthy();
        expect(iframe.getAttribute("src")).toBe("url");
        expect(iframe.getAttribute("height")).toBe("100px");
        expect(iframe.getAttribute("width")).toBe("150px");
    });

    test("Реклама автоматически закрывается после 10 секунд", async () => {
        const banner = createBanner({ params: { isTvRemote: true } });
        await banner.run();
        let rootElement = document.getElementById("root");
        expect(rootElement).toBeTruthy();

        triggerOnLoadIframe();

        jest.advanceTimersByTime(5 * 1000);

        rootElement = document.getElementById("root");
        expect(rootElement).toBeTruthy();

        jest.advanceTimersByTime(5 * 1000);

        jest.advanceTimersByTime(1000);

        rootElement = document.getElementById("root");
        expect(rootElement).toBeFalsy();
    });

    test('Реклама закрывается при нажатии на кнопку "назад" на пульте', async () => {
        const banner = createBanner({ params: { isTvRemote: true } });
        await banner.run();
        let rootElement = document.getElementById("root");
        expect(rootElement).toBeTruthy();

        triggerOnLoadIframe();

        history.back();

        jest.advanceTimersByTime(1000);

        rootElement = document.getElementById("root");
        expect(rootElement).toBeFalsy();
    });

    test("Таймер меняется корректно раз в секунду", async () => {
        const banner = createBanner({ params: { isTvRemote: true } });
        await banner.run();

        const timerElement = document.getElementById("plasma-ad-time-to-close");

        triggerOnLoadIframe();

        for (let i = 9; i > 0; i--) {
            jest.advanceTimersByTime(1 * 1000);
            expect(timerElement.innerText).toBe(i.toString());
        }
    });
});
