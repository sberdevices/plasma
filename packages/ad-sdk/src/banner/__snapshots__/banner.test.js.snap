// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`banner.js - Banner class Совпадает снапшот 1`] = `
<body>
  <div
    id="root"
  >
    <div
      style="position: absolute; top: 0px; left: 0px; width: 100%; height: 100%; display: flex;"
    >
      <div
        style="margin: auto; flex-shrink: 0; flex-grow: 0; max-width: 100vw; max-height: 100vh; padding: 5px; box-sizing: border-box; display: flex; flex-direction: column; align-items: strength;"
      >
        <div
          style="display: flex; align-items: center; box-sizing: border-box; padding: 5px; flex-grow: 0; flex-shrink: 0;"
        >
          <div
            style="color: white; flex-grow: 1; flex-shrink: 1;"
          >
            Реклама закроется через 
            <span
              id="plasma-ad-time-to-close"
            >
              10
            </span>
            с
          </div>
          <button
            style="background: none; padding: 0px; margin: 0px; position: relative; flex-grow: 0; flex-shrink: 0;"
          >
            <svg
              color="inherit"
              fill="none"
              height="40px"
              style="display: block;"
              viewBox="0 0 24 24"
              width="40px"
            >
              <path
                d="M12 2c5.522 0 10 4.477 10 10s-4.478 10-10 10S2 17.523 2 12 6.478 2 12 2zM8.707 7.292a1 1 0 00-1.414 1.415L10.586 12l-3.293 3.293a.999.999 0 00-.083 1.32l.083.094a.997.997 0 001.414 0L12 13.414l3.293 3.293a.996.996 0 00.58.285L16 17a.999.999 0 00.707-1.707L13.414 12l3.293-3.293a1 1 0 00.083-1.32l-.083-.095a1 1 0 00-1.414 0L12 10.585 8.707 7.292z"
                fill="white"
              />
            </svg>
          </button>
        </div>
        <a
          href="https://www.google.com/"
          id="plasma-ad-banner-link"
          style="flex-grow: 1; flex-shrink: 1;"
        >
          <iframe
            height="100px"
            src="url"
            width="150px"
          />
        </a>
      </div>
    </div>
  </div>
</body>
`;
